pkg/crawl/finding.go

```diff
@@ -0,0 +1,117 @@
+
+package crawl
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+const (
+	UnknownVersion = "unknown"
+
+	NothingDetected                Finding = 0
+	JndiLookupClassName            Finding = 1 << iota
+	JndiLookupClassPackageAndName  Finding = 1 << iota
+	JndiManagerClassName           Finding = 1 << iota
+	JarName                        Finding = 1 << iota
+	JarNameInsideArchive           Finding = 1 << iota
+	JndiManagerClassPackageAndName Finding = 1 << iota
+	JarFileObfuscated              Finding = 1 << iota
+	ClassBytecodePartialMatch      Finding = 1 << iota
+	ClassBytecodeInstructionMd5    Finding = 1 << iota
+	ClassFileMd5                   Finding = 1 << iota
+)
+
+var (
+	// vulnerableFindingStrings contains only the Finding values that are considered vulnerable.
+	vulnerableFindingStrings = map[Finding]string{
+		JndiLookupClassName:            "JndiLookupClassName",
+		JndiLookupClassPackageAndName:  "JndiLookupClassPackageAndName",
+		JndiManagerClassName:           "JndiManagerClassName",
+		JarName:                        "JarName",
+		JarNameInsideArchive:           "JarNameInsideArchive",
+		JndiManagerClassPackageAndName: "JndiManagerClassPackageAndName",
+		JarFileObfuscated:              "JarFileObfuscated",
+		ClassBytecodePartialMatch:      "ClassBytecodePartialMatch",
+		ClassBytecodeInstructionMd5:    "ClassBytecodeInstructionMd5",
+		ClassFileMd5:                   "ClassFileMd5",
+	}
+
+	stringFindings           = invertedFindingStringMap()
+	lowercasedStringFindings = lowerCasedInvertedFindingStringMap()
+)
+
+type Finding int
+type Versions map[string]struct{}
+
+// FindingOf creates a finding from a string, returning an error if a corresponding finding does not exist.
+// Conversion is case-insensitive.
+func FindingOf(v string) (Finding, error) {
+	finding, ok := lowercasedStringFindings[strings.ToLower(v)]
+	if !ok {
+		return 0, fmt.Errorf("invalid finding-match %s, supported values are %s", v, strings.Join(SupportedVulnerableFindingValues(), ", "))
+	}
+	return finding, nil
+}
+
+func SupportedVulnerableFindingValues() []string {
+	var supportedValues []string
+	for s := range stringFindings {
+		supportedValues = append(supportedValues, s)
+	}
+	sort.Strings(supportedValues)
+	return supportedValues
+}
+
+func lowerCasedInvertedFindingStringMap() map[string]Finding {
+	out := make(map[string]Finding)
+	for s, finding := range invertedFindingStringMap() {
+		out[strings.ToLower(s)] = finding
+	}
+	return out
+}
+
+func invertedFindingStringMap() map[string]Finding {
+	out := make(map[string]Finding)
+	for finding, s := range vulnerableFindingStrings {
+		_, exists := out[s]
+		if exists {
+			panic(fmt.Sprintf("finding already defined when inverting finding strings map: %s", s))
+		}
+		out[s] = finding
+	}
+	return out
+}
+
+func (f Finding) String() string {
+	var out []string
+	// for all non-zero bits, append string for finding if it exists
+	for i := 0; f > 0; i, f = i+1, f>>1 {
```

Comment:
This logic is a bit hard to parse/read to me -- personally, I would find it easier to follow if it was a for loop of the form:

```
for i := JndiLookupClassName /* or NothingDetected? */; i != (valueAfterClassFileMd5); i>>1 {
    // logic
}
```

The above construction would make it clearer to me that we're iterating through each valid "Finding" type and appending based on that. However, it would require defining a new terminal `Finding` type (so that there is an end point for the iteration).

I don't feel strongly enough about this to assert that the change must be made, but wanted to provide the feedback.